src/compose.ts

```diff
@@ -346,3 +366,99 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+/**
+ * Parses a composition and returns a list of image descriptors that can
+ * be used to pull or build a service image. The given composition version
+ * must be equal to `DEFAULT_SCHEMA_VERSION`, or an exception is thrown.
+ * Normalise the composition before passing it to this function.
+ */
+async function expandContext(
+	c: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	const cachedEnvironmentFiles: Dict<Dict<string>> = {};
+
+	// Apply all read env_files content to the services referncing the env_files
+	for (const service of Object.values(c.services)) {
```

Comment:
My personal preference would be split this single loop into two operations.

1. Loading and caching env files
1. Updating the service's environment

Reply:
As the caching is not the main purpose of the env_file parsing, I like it to be within the main logic.
Anyway, I'm fine with both. 

Same applies here, I'd like to have @dfunckt in the loop, as he suggested to combine all necessary loops into one.

Reply:
Since service can have 1 or many env_files and they are shared, env files are being read and parsed incrementally and then "cached" in case the next services require them. 

If its separated then first all the env files are loaded and parsed and then for each service the environment is updated based on which files they reference. You don't really need to call it a "cache" anymore any more that any variable is a cache.

Reply:
I'll stick on the 1 loop approach, as it's more intuitive to read, in contrast to follow up multiple loops that decouple the logic of caching / prereading the env files

Reply:
Ok all good up to you. Just to expand on my thinking for the intellectual persuit. 

Decoupling logic is usually a good thing. The reason you consider it caching or prereading is because the logic is not decoupled. It becomes a caching when done incrementally. caching is a more complex concept than reading, it's an optimisation strategy to prevent reading the same file multiple times. So caching increases complexity to increase efficiency.

However if all Env files are pulled out of the data structure and read at once that negates the need for a cache type optimisation and can just be considered reading. Loops on arrays less than 10k items are insanely cheap and would likely be jitted away.

Instead of thinking of it as 2 loops you could think of it as 2 functions something like readEnvFiles and assignEnvFileToEnvironment I think the functionality would be easier to understand without having to read every line.

Reply:
@codewithcheese Thanks for expanding you thoughts, and I follow all on them. I like the splitting of the functionality into functions. As I've named it expandContext, the context could increase and then readingEnvFiles is one part of the context and should reside in a separate function.
I'll apply this rationell and will recommit, thanks a lot.